### DISCUSSION TITLE
VSCode Integration - Vertical and Horizontal Scrolling

### DISCUSSION THREAD

**Discussion by jmenzies12** [eclipse-glsp/glsp#1447] [Q&A] [answered]
Hey there,

I noticed in the VSCode Integration, that scrolling the mouse wheel (or relevant trackpad function) zooms out instead of scrolling the canvas vertically. Also, dragging a node to the edge of the canvas doesn't scroll the canvas. 

Is this intended behavior? And would it be possible for us to override it?

Kind Regards
Jonathan

**Comment by martin-fleck-at** [comment, marked as answer]
Hi @jmenzies12! Please excuse the late reply around the holidays.

Yes, if you are using the [viewport module](https://github.com/eclipse-glsp/glsp-client/blob/61906cf0bcdf68981ead369a34e0112a857f1161/packages/client/src/features/viewport/viewport-modules.ts#L36C14-L36C28) which is added by default, it binds the [ZoomListener](https://github.com/eclipse-sprotty/sprotty/blob/079a6913e5a44b0c4a8efecd0020680bb2d67461/packages/sprotty/src/features/viewport/zoom.ts#L39) that zooms in when using the mouse wheel. By pressing ALT you can switch to a panning mode for moving up and down. If you want to change that, you'd need to rebind the ZoomMouseListener and add your own custom behavior.

As for the shifting of the canvas when dragging a node to the edge, that is indeed not implemented. However, we have that behavior when moving an element through the arrow keys, see the [MoveElementHandler](https://github.com/eclipse-glsp/glsp-client/blob/61906cf0bcdf68981ead369a34e0112a857f1161/packages/client/src/features/accessibility/move-zoom/move-handler.ts#L200-L202). I think something like that could be integrated into the move tool when dragging. As before you need to rebind the respective listener to adapt the behavior here.

I hope this helps you with what you are trying to achieve.
